server/src/main/java/org/opensearch/cluster/store/RemoteClusterStateService.java

```diff
@@ -0,0 +1,247 @@
+            allUploadedIndexMetadata.put(indexMetadata.getIndex().getName(), uploadedIndexMetadata);
+        }
+        return uploadMarker(clusterState, allUploadedIndexMetadata);
+    }
+
+    @Nullable
+    public ClusterMetadataMarker writeIncrementalMetadata(
+        ClusterState previousClusterState,
+        ClusterState clusterState,
+        ClusterMetadataMarker previousMarker
+    ) throws IOException {
+        if (clusterState.nodes().isLocalNodeElectedClusterManager() == false) {
+            logger.error("Local node is not elected cluster manager. Exiting");
+            return null;
+        }
+        assert previousClusterState.metadata().coordinationMetadata().term() == clusterState.metadata().coordinationMetadata().term();
+        final Map<String, Long> indexMetadataVersionByName = new HashMap<>();
+        for (final IndexMetadata indexMetadata : previousClusterState.metadata().indices().values()) {
+            indexMetadataVersionByName.putIfAbsent(indexMetadata.getIndex().getName(), indexMetadata.getVersion());
+        }
+
+        int numIndicesUpdated = 0;
+        int numIndicesUnchanged = 0;
+        final Map<String, ClusterMetadataMarker.UploadedIndexMetadata> allUploadedIndexMetadata = new HashMap<>(
+            previousMarker.getIndices()
+        );
+        for (final IndexMetadata indexMetadata : clusterState.metadata().indices().values()) {
+            final Long previousVersion = indexMetadataVersionByName.get(indexMetadata.getIndex().getName());
+            if (previousVersion == null || indexMetadata.getVersion() != previousVersion) {
+                logger.trace(
+                    "updating metadata for [{}], changing version from [{}] to [{}]",
+                    indexMetadata.getIndex(),
+                    previousVersion,
+                    indexMetadata.getVersion()
+                );
+                numIndicesUpdated++;
+                final String indexMetadataKey = writeIndexMetadata(
+                    clusterState.getClusterName().value(),
+                    clusterState.getMetadata().clusterUUID(),
+                    indexMetadata,
+                    indexMetadataFileName(indexMetadata)
+                );
+                final UploadedIndexMetadata uploadedIndexMetadata = new UploadedIndexMetadata(
+                    indexMetadata.getIndex().getName(),
+                    indexMetadata.getIndexUUID(),
+                    indexMetadataKey
+                );
+                allUploadedIndexMetadata.put(indexMetadata.getIndex().getName(), uploadedIndexMetadata);
+            } else {
+                numIndicesUnchanged++;
+            }
+            indexMetadataVersionByName.remove(indexMetadata.getIndex().getName());
+        }
+
+        for (String removedIndexName : indexMetadataVersionByName.keySet()) {
+            allUploadedIndexMetadata.remove(removedIndexName);
+        }
+        return uploadMarker(clusterState, allUploadedIndexMetadata);
```

Comment:
What is the expected state of `allUploadedIndexMetadata`? Should it contain metadata of all the existing indices in the cluster?